### PR TITLE
added runner for Fairseq

### PR DESCRIPTION
**Pull request by taylanbil**
docstring contains a bash script as an example (transformer).

### CODE REVIEW

File: fairseq_train_tpu.py [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+"""
+
+import argparse
+import sys
+import os
+import math
+import collections
+from datetime import datetime
+
+root_folder = os.path.dirname(os.path.abspath(sys.argv[0]))
```
[the remark this answers is not in the document]
Comment (taylanbil) [reply fairseq_train_tpu.py]: agreed, will update PR w/ new utils file that has this.

File: fairseq_train_tpu.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+                       left_pad=False,
+                       move_eos_to_beginning=False):
+  # Copied over from fairseq.data_utils, and modified so that num_columns
+  # in the output tensor is not too variable.
+
+  # correcting columns
+  global PAD_TO_LENGTH
+  size = max(v.size(0) for v in values)
+  if size > PAD_TO_LENGTH:
+    print(
```
[the remark this answers is not in the document]
Comment (taylanbil) [reply fairseq_train_tpu.py]: will do, thanks

File: fairseq_train_tpu.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+  train_meter.stop()
+  print('| done training in {:.1f} seconds'.format(train_meter.sum))
+
+
+def count_compiles():
+  metricsreport = torch_xla._XLAC._xla_metrics_report().split('\n')
+  for i, line in enumerate(metricsreport):
+    if line.endswith('CompileTime'):
+      return int(''.join([c for c in metricsreport[i + 1] if c.isdigit()]))
```
[the remark this answers is not in the document]
Comment (taylanbil) [reply fairseq_train_tpu.py]: oh I was going to delete this function altogether. This is for me to keep track of compiles per step while debugging.

File: fairseq_train_tpu.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        args,
+        itr,
+        epoch_itr.epoch,
+        prefix='valid on \'{}\' subset'.format(subset),
+        no_progress_bar='simple')
+    stats_per_device = model_parallel(valid_loop_fn, progress)
+    valid_losses = [stats['loss'].avg for stats in stats_per_device]
+    print('validation stats on subset "{}" - {}'.format(subset, now()))
+    for stats in stats_per_device:
+      print(now())
```
[the remark this answers is not in the document]
Comment (taylanbil) [reply fairseq_train_tpu.py]: sorry, leftover from debugging.

**Comment by taylanbil** [comment]
> A more general question.
> I guess the bulk of this file comes from one of their "runners", right?
> If yes, are there any reads by which we can split our changes into up-streamable ones, and XLA ones?

Yes, bulk of this file is from https://github.com/pytorch/fairseq/blob/master/train.py
However, I have not been able to isolate the upstreamable changes any further than this and maintain functionalities offered by fairseq's `task`, `trainer`, `progress` objects etc. I did upstream a few changes, but 

Part of it comes down to the way train.py is coded up, needing to inject our train and valid loops, and not reporting loss so often etc led to this file.

File: fairseq_train_tpu.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+if __name__ == '__main__':
+  # override certain args so that we use XLA parallelism instead of torch.
+  FLAGS = parse_args()
+  if FLAGS.use_gpu:
+    data_utils.collate_tokens = collate_tokens_gpu
+    fairseq_train.cli_main()
+  else:
+    BATCH_SIZE = FLAGS.max_sentences
+    DEVICES = xm.get_xla_supported_devices(max_devices=FLAGS.num_cores)
```
[the remark this answers is not in the document]
Comment (taylanbil) [reply fairseq_train_tpu.py]: deleted BATCH_SIZE, it wasn't used heavily anyway
moved DEVICES inside the `main_tpu` function

I do need PAD_TO_LENGTH as a global in order to access it inside `collate_tokens`, and not having to add it to fairseq's custom data classes so it trickles down to that function.

**Comment by taylanbil** [comment]
tested, current version works

**Comment by taylanbil** [comment]
btw, this PR also points the fairseq dependency to the correct commit
